data_steward/curation_logging/curation_gae_handler.py

```diff
@@ -0,0 +1,487 @@
+            'userAgent': self._request_agent,
+            'host': self._request_host,
+            'ip': self._request_remote_addr,
+            'responseSize': self._response_size,
+            'status': self._response_status_code,
+            'requestId': self._request_log_id,
+            'traceId': self._trace,
+            'versionId': os.environ.get('GAE_VERSION', 'devel'),
+            'urlMapEntry': 'validation.main.app'
+        }
+
+        if self._request_taskname:
+            proto_payload_args['taskName'] = self._request_taskname
+            proto_payload_args['taskQueueName'] = self._request_queue
+
+        if self._first_log_ts:
+            if self._first_log_ts:
+                total_time = datetime.utcnow() - self._first_log_ts
+            else:
+                total_time = 0
+            proto_payload_args['latency'] = '{0}.{1}s'.format(total_time.seconds, total_time.microseconds)
+
+        proto_payload_pb2 = setup_proto_payload(
+            lines,
+            self.log_completion_status,
+            **proto_payload_args
+        )
+
+        log_entry_pb2_args['proto_payload'] = proto_payload_pb2
+
+        # https://cloud.google.com/logging/docs/reference/v2/rpc/google.logging.v2#google.logging.v2.LogEntry
+        log_entry_pb2 = gcp_logging_v2.types.log_entry_pb2.LogEntry(**log_entry_pb2_args)
+
+        self._logging_client.write_log_entries([log_entry_pb2],
+                                               log_name=LOG_NAME_TEMPLATE.
+                                               format(project_id=app_identity.get_application_id()))
+
+
+def get_gcp_logger() -> GCPStackDriverLogger:
+    """
+    Return the GCPStackDriverLogger object for this thread.
+    :return: GCPStackDriverLogger object
+    """
+    if hasattr(_thread_store, 'logger'):
+        _logger = getattr(_thread_store, 'logger')
+        return _logger
+
+    # We may need to initialize the logger for this thread.
+    _logger = GCPStackDriverLogger()
+    setattr(_thread_store, 'logger', _logger)
+    return _logger
+
+
+class GCPLoggingHandler(logging.Handler):
+
+    def emit(self, record: logging.LogRecord):
+        """
+        Capture and store a log event record.
+        :param record: Python log record
+        """
+        _logger = get_gcp_logger()
+        if _logger:
+            _logger.log_event(record)
+            return
+
+        line = setup_log_line(record)
+        print(line)
+
+
+def initialize_logging(log_level=logging.INFO):
+    """
+    Setup GCP Stack Driver logging if we are running in App Engine.
+    :param log_level: Log level to use.
+    """
+    # Configure root logger
+    root_logger = logging.getLogger()
+    root_logger.setLevel(log_level)
+    # Configure StackDriver logging handler
+    log_handler = GCPLoggingHandler()
+    log_handler.setLevel(log_level)
+    # Add StackDriver logging handler to root logger.
+    root_logger.addHandler(log_handler)
+
+
+initialize_logging()
```

Comment:
@ChaoPang I wonder if we should be concerned about this call- it seems it would cause side effects on import of the `curation_gae_handler` module. If so, it may lead to multiple initialization of the root logger unintentionally and it is possible this may be related to the bug we've seen previously.